src/sos.h

```diff
@@ -0,0 +1,169 @@
+//
+//  sos.h
+//  sos
+//
+//  Created by Pavan Kumar Sunkara on 10/01/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+
+#ifndef SOS_H
+#define SOS_H
+
+#include <string>
+#include <vector>
+#include <map>
+#include <ostream>
+#include <algorithm>
+#include <memory>
+
+namespace sos {
+
+    /** Forward Declaration */
+    class Base;
+
+    /** Key-value pairs (object) */
+    typedef std::map<std::string, Base> KeyValues;
+
+    /** List of keys for object (for deterministic output) */
+    typedef std::vector<std::string> Keys;
+
+    /** Array of values */
+    typedef std::vector<Base> Bases;
+
+    /** Value base structure */
+    class Base {
+    public:
+
+        /** Enum for Type of the value */
+        enum Type {
+            NullType = 1, // Null
+            StringType,   // A string
+            NumberType,   // A number
+            BooleanType,  // A boolean value
+            ArrayType,    // An array
+            ObjectType,   // An object
+            UndefinedType = 0
+        };
+
+        /** Type of the value */
+        Type type;
+
+        /** String value */
+        std::string string;
+
+        /** Number value */
+        double number;
```

Comment:
Actually, double handles both `int` and `float` when printing. See the test cases.